### ISSUE TITLE
NeurIPS '19 | Convolution with even-sized kernels and symmetric padding.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#80]
Wu S, Wang G, Tang P, et al. [Convolution with even-sized kernels and symmetric padding](https://proceedings.neurips.cc/paper/2019/hash/2afe4567e1bf64d32a5527244d104cea-Abstract.html).

**Comment by NorbertZheng** [comment]
## Overview
This paper explores the influence of even kernel size convolution on the network, and the results show that
 - even convolution is not as good as odd convolution in terms of results.

The article concludes from experiments and principles that the reason why even-numbered convolutions are worse is because
 - even-numbered convolutions will shift the feature map, that is, "the shift problem", which will cause the learned features to not have stronger representational ability.

This paper puts forward the information erosion hypothesis, and believes that the center of the odd convolution is symmetrical, while the even convolution has no symmetric point when it is realized. This will cause the information used by the convolution to be realized in all directions, but only in the upper left or other directions (depending on the In the specific implementation), so the whole will cause the feature map to shift in one direction. In order to solve this problem, the article
 - proposes a symmetric padding method to make up for the losses caused by all directions,

and the results are obvious.

**Comment by NorbertZheng** [comment]
## The Shift Problem
The convolution of odd kernel size is easy to implement. Calculate the feature value and weight sum of the current position and its eight neighborhood directions at the corresponding position, which is a value corresponding to the output position. How to realize the convolution of even kernel size? ? In tensorflow, the convolution of an even kernel size such as 2×2 convolution is obtained by
 - multiplying and summing the current point and its upper left, upper, and left points with the corresponding weights.

Because of this, in the implementation process , the receptive field of the output feature map is actually flawed. It can only correspond to the upper left area. The problem exposed after multi-layer convolution is the offset of the feature map.

![image](https://user-images.githubusercontent.com/37235023/232366336-4453adbb-472c-4aa5-9d9f-1fae67cfea74.png)

As shown in the figure above, the first line is the result of conv2x2 without using the method padding in this paper, and the second line is the result of the method in this paper.

The relationship between the output feature map and the input feature map can be expressed as:
![image](https://user-images.githubusercontent.com/37235023/232366439-855aff7e-cd9d-4d34-a4aa-ba6b436227fc.png)

Where $p$ represents the position coordinates, $F^{i}$ and $F^{o}$ represent the input feature and output feature respectively, $\delta$ represents the position in the convolution kernel, and $\omega_{i}$ represents the weight of the convolution kernel. For odd kernel size convolution, where $R$ can be written as:
![image](https://user-images.githubusercontent.com/37235023/232366871-fb6a3be8-bca4-4093-a2b5-a32c4a78ab83.png)

Taking the center of the kernel as the origin, the offset value in each direction is the above representation. For odd convolution, it is clear that the above formula is centrally symmetric, while for even kernel size convolution, the corresponding $R$ is defined as:
![image](https://user-images.githubusercontent.com/37235023/232366967-ea97e8ef-8aad-43f3-b0d1-2cf5bb07578a.png)

The above formula does not use information in all directions, and the convolution kernel is not centrally symmetrical.

**Comment by NorbertZheng** [comment]
## The Information Erosion Hypothesis
For the input position $p$, the corresponding position after $n$ times of even convolution is:
![image](https://user-images.githubusercontent.com/37235023/232367247-1e315e0d-917c-44dd-aead-fcfd11e941d5.png)

Therefore, the deeper the network, the more serious the shift phenomenon.

In order to illustrate the eroding effect of even-numbered convolutions on information, the article defines the L1 norm of the feature as a measure of the amount of information in the feature map.
![image](https://user-images.githubusercontent.com/37235023/232367324-1e751e92-c9b2-4d36-baa9-27f189b31dba.png)

Based on this definition, the article experimented with the amount of information of different kernel sizes, and obtained the results shown in the figure:
![image](https://user-images.githubusercontent.com/37235023/232367372-40582811-b999-4092-9638-2c79904d997a.png)

It can be seen that the overall information content of C3 and C5 is much slower than that of C2 and C4.

**Comment by NorbertZheng** [comment]
## Symmetric Padding
In order to solve the impact brought by shift, this paper proposes a padding method, the specific operation is shown in the figure:
![image](https://user-images.githubusercontent.com/37235023/232367475-d779b65b-2da2-429c-a877-2aaf54e3ed3a.png)

That is, first divide the feature map into four groups, each group performs padding in different directions as shown in the figure, and finally conv2x2 without padding.

The advantage of this is that certain channels of the network can use information in a specific direction. From a macro perspective, the network can use information in all directions, which alleviates the problems caused by shift to a certain extent.

**Comment by NorbertZheng** [comment]
## Coding
```pytthon
import torch
import torch.nn as nn
import torch.nn.functional as F

class SpConv2d(nn.Module):
    def __init__(self,in_channels,out_channels,kernel_size,stride,padding,*args,**kwargs):
        super(SpConv2d,self).__init__()
        self.conv = nn.Conv2d(in_channels,out_channels,kernel_size,stride,padding)
    def forward(self,x):
        n,c,h,w = x.size()
        assert c % 4 == 0
        x1 = x[:,:c//4,:,:]
        x2 = x[:,c//4:c//2,:,:]
        x3 = x[:,c//2:c//4*3,:,:]
        x4 = x[:,c//4*3:c,:,:]
        x1 = nn.functional.pad(x1,(1,0,1,0),mode = "constant",value = 0) # left top
        x2 = nn.functional.pad(x2,(0,1,1,0),mode = "constant",value = 0) # right top
        x3 = nn.functional.pad(x3,(1,0,0,1),mode = "constant",value = 0) # left bottom
        x4 = nn.functional.pad(x4,(0,1,0,1),mode = "constant",value = 0) # right bottom
        x = torch.cat([x1,x2,x3,x4],dim = 1)
        return self.conv(x)

class Net(nn.Module):
    def __init__(self):
        super(Net,self).__init__()
        self.conv = SpConv2d(4,16,2,1,0)
    def forward(self,x):
        return self.conv(x)
if __name__ == "__main__":
    x = torch.randn(2,4,14,14)
    net = Net()
    print(net(x))
```

**Comment by NorbertZheng** [comment]
## Conclusion
Compact convolutional neural networks gain efficiency mainly through depthwise convolutions, expanded channels and complex topologies, which contrarily aggravate the training process. Besides, 3x3 kernels dominate the spatial representation in these models, whereas even-sized kernels (2x2, 4x4) are rarely adopted. In this work, we quantify the shift problem occurs in even-sized kernel convolutions by an information erosion hypothesis, and eliminate it by proposing symmetric padding on four sides of the feature maps (C2sp, C4sp). Symmetric padding releases the generalization capabilities of even-sized kernels at little computational cost, making them outperform 3x3 kernels in image classification and generation tasks. Moreover, C2sp obtains comparable accuracy to emerging compact models with much less memory and time consumption during training. Symmetric padding coupled with even-sized convolutions can be neatly implemented into existing frameworks, providing effective elements for architecture designs, especially on online and continual learning occasions where training efforts are emphasized.
